src/main/io/displayport_hott.c

```diff
@@ -0,0 +1,187 @@
+        return 0;
+
+    hottTextmodeWriteChar(col, row, c);
+    return 0;
+}
+
+static int hottWriteString(displayPort_t *displayPort, uint8_t col, uint8_t row, const char *s)
+{
+    while (*s) {
+        hottWriteChar(displayPort,  col++, row, *(s++));
+    }
+    return 0;
+}
+
+static int hottClearScreen(displayPort_t *displayPort)
+{
+    for (int row = 0; row < displayPort->rows; row++) {
+        for (int col= 0; col < displayPort->cols; col++) {
+            hottWriteChar(displayPort, col, row, ' ');
+        }
+    }
+    return 0;
+}
+
+static bool hottIsTransferInProgress(const displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+    return false;
+}
+
+static int hottHeartbeat(displayPort_t *displayPort)
+{
+    if (!hottTextmodeIsAlive())
+        cmsMenuExit(displayPort, (void*)CMS_EXIT_SAVE);
+
+    return 0;
+}
+
+static void hottResync(displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+}
+
+static uint32_t hottTxBytesFree(const displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+    return UINT32_MAX;
+}
+
+static int hottGrab(displayPort_t *displayPort)
+{
+    hottTextmodeGrab();
+    return displayPort->grabCount = 1;
+}
+
+static int hottRelease(displayPort_t *displayPort)
+{
+    int cnt = displayPort->grabCount = 0;
+    hottClearScreen(displayPort);
+    hottTextmodeExit();
+    return cnt;
+}
+
+static const displayPortVTable_t hottVTable = {
+    .grab = hottGrab,
+    .release = hottRelease,
+    .clearScreen = hottClearScreen,
+    .drawScreen = hottDrawScreen,
+    .screenSize = hottScreenSize,
+    .writeString = hottWriteString,
+    .writeChar = hottWriteChar,
+    .isTransferInProgress = hottIsTransferInProgress,
+    .heartbeat = hottHeartbeat,
+    .resync = hottResync,
+    .txBytesFree = hottTxBytesFree
+};
+
+displayPort_t *displayPortHottInit()
+{
+    hottDisplayPort.device = NULL;
+    displayInit(&hottDisplayPort, &hottVTable);
+    hottDisplayPort.rows = HOTT_TEXTMODE_DISPLAY_ROWS + ROW_CORRECTION * 2;
+    hottDisplayPort.cols = HOTT_TEXTMODE_DISPLAY_COLUMNS + COL_CORRECTION * 2;
+    return &hottDisplayPort;
+}
+
+void hottCmsOpen()
+{
+    if (!cmsInMenu) {
+        cmsDisplayPortSelect(&hottDisplayPort);
+        cmsMenuOpen();
+    }
+}
+
+void hottSetCmsKey(uint8_t hottKey, bool esc)
+{
+    switch (hottKey) {
+        case HOTTV4_BUTTON_DEC:
+            cmsSetExternKey(CMS_KEY_UP);
+            break;
+        case HOTTV4_BUTTON_INC:
+            cmsSetExternKey(CMS_KEY_DOWN);
+            break;
+        case HOTTV4_BUTTON_SET:
+            if (cmsInMenu)
+                cmsMenuExit(pCurrentDisplay, (void*)CMS_EXIT_SAVE);
+            cmsSetExternKey(CMS_KEY_NONE);
+            break;
+        case HOTTV4_BUTTON_NEXT:
+            cmsSetExternKey(CMS_KEY_RIGHT);
+            break;
+        case HOTTV4_BUTTON_PREV:
+            cmsSetExternKey(CMS_KEY_LEFT);
+            if (esc)
+                cmsMenuOpen();
```

Comment:
there are 2 way to enter menu - sticks and key, kut key processing is mixed together... this makes code harder to read/maintain ...

Reply:
Not really, yes you can select the HoTT CMS menu by sticks, but only when you access the cms by key its get visible. This code makes sure the CMS is open until textmode is closed.  
Perhaps the argument name should be keepCmsOpen.